migration/migrator/migrations/course/20210610100802_course_materials_json_to_db.py

```diff
@@ -0,0 +1,121 @@
+"""Migration for a given Submitty course database."""
+
+import json
+from pathlib import Path
+from sqlalchemy import insert
+
+def up(config, database, semester, course):
```

Comment:
It looks to me like this `up` migration would fail if we tried to run this twice (i.e. `migrate`, `rollback`, `migrate`). Your best bet might be to replace your `inserts` with [`upserts`](https://wiki.postgresql.org/wiki/UPSERT)?

Reply:
Yeah this is a good point. I just updated the queries to add a on conflict to do upserts. Can you take a look at the new statements?

Reply:
It looks like you've re-added dropping the tables in the down part of the migration that it would render the upsert obsolete and you can just do a regular insert.

Reply:
The plan is to not have the down migration. I had those for testing only. I do plan on sticking with the upserts.

Reply:
Given the nature of the change, I think dropping the tables is fine. Having thought about this more, this is an instance where we want to do a "true" rollback, where within the migration you:

1. create the tables
2. read the JSON into the tables

and so then in the rollback, you'd:

1. read the tables into the JSON file
2. drop the tables

as the source of truth shifts from JSON file to tables within this PR, and vice versa if we were to rollback.

Reply:
I vote for not wasting time implementing the (nontrivial) down migration.  I vote for an empty down migration.  This is a big PR and a big change, but this change is LONG overdue.  We will not be looking backward once we switch to the database.
If we do make a mistake, we might lose metadata if we have to revert and bugfix and reapply, but I am ok with the potential data loss.  Importantly, we won't lose any of the course materials files.